[GEOS-8034] Fixed ysld color conversion

hwbllmnn opened the pull request:
This fixes GEOS-8034. I'm unsure why the literals were using java.awt.Color objects before, even the tests were checking for that. I tested with Geoserver 2.11 and ysld styles work just fine with this fix. I'll also provide a fix for the master branch shortly.

aaime commented:
Hey, I'm not sure about the merits of the pull request (unfamiliar with ysld and the code backing it), but from a procedure point of view, every change needs to land on master first, and then be backported. At this point I'd probably wait for the module maintainers review.

hwbllmnn commented:
Hi, sure, see #1649 . Am I at least correct in the assumption that OGC literals for color values should be created with a hex color string as their value? Or can/should they be created with java.awt.Color objects as their value?

aaime commented:
The rendering subsystem should be happy with java.awt.Color as well, as far as I know. Maybe if you're trying to covert ysld to SLD, it can cause issues in the code dumping down the style.

tbarsballe commented on modules/extension/ysld/src/main/java/org/geotools/ysld/parse/Util.java:
```diff
@@ -218,7 +218,13 @@ public static Expression color(Object value, Factory factory) {
         if (value != null)
             value = value.toString();
 
-        return color != null ? factory.filter.literal(color) : expression((String) value, factory);
```
Rather than changing the behavior within YSLD, you should instead update the GeoTools [color converter factory] (https://github.com/geotools/geotools/blob/master/modules/library/main/src/main/java/org/geotools/util/ColorConverterFactory.java) to properly handle these values.

tbarsballe commented on modules/extension/ysld/src/test/java/org/geotools/ysld/parse/UtilTest.java:
```diff
@@ -29,12 +27,12 @@
+        assertEquals("#00FFFF",
                 ((Literal) Util.color("rgb(0, 255, 255)", factory)).getValue());
-        assertEquals(new Color(0xFF00FF), ((Literal) Util.color("f0f", factory)).getValue());
+        assertEquals("#FF00FF", ((Literal) Util.color("f0f", factory)).getValue());
```
These tests should still work as-is. If they don't, you may have broken some other behaviour in YSLD.

tbarsballe commented on modules/extension/ysld/src/test/java/org/geotools/ysld/parse/YsldParseTest.java:
```diff
@@ -1240,13 +1235,13 @@ public void testColourMapEmpty() throws Exception {
-        assertThat(colour3, is(Color.BLUE));
+        assertThat(colour1, is("#FF0000"));
+        assertThat(colour2, is("#00FF00"));
+        assertThat(colour3, is("#0000FF"));
```
These tests should still work as-is. If they don't, you may have broken some other behaviour in YSLD.

hwbllmnn commented:
I think I found the real culprit, the YsdlEncodeHandler converted all objects to String (and tries to convert back to number if possible). I added an exception in the case of a literal with a single Object, in that case it is returned directly. I also added a test for this.

Do you think that's the proper solution?

hwbllmnn commented:
Backported requested changes from master PR.

tbarsballe commented:
#1649 was merged to master, and manually backported to 18.x and 17.x
